tools/Voron.Recovery/RecoveredDatabaseCreator.cs

```diff
@@ -0,0 +1,433 @@
+using System.Globalization;
+using System.IO;
+using Raven.Client;
+using Raven.Client.Documents.Operations.Attachments;
+using Raven.Client.Documents.Operations.Counters;
+using Raven.Client.Documents.Operations.Revisions;
+using Raven.Client.Documents.Session;
+using Raven.Client.Documents.Smuggler;
+using Raven.Client.Json;
+using Raven.Server.Documents;
+using Raven.Server.ServerWide;
+using Raven.Server.Smuggler.Documents;
+using Raven.Server.Smuggler.Documents.Data;
+using Sparrow.Json;
+using Raven.Client.Extensions;
+using Raven.Client.Util;
+using Raven.Server.ServerWide.Context;
+using Sparrow.Json.Parsing;
+using Sparrow.Logging;
+using Sparrow.Server;
+using Sparrow.Threading;
+using Voron.Data;
+
+namespace Voron.Recovery
+{
+    public class RecoveredDatabaseCreator : IDisposable
+    {
+        public static RecoveredDatabaseCreator RecoveredDbTools(DocumentDatabase documentDatabase, string recoverySession, Logger logger) => new RecoveredDatabaseCreator(documentDatabase, recoverySession, logger);
+
+        private readonly DatabaseDestination _databaseDestination;
+        private readonly SmugglerResult _results;
+        private readonly DocumentsOperationContext _context;
+        private readonly DocumentDatabase _database;
+        private readonly IDisposable _contextDisposal;
+        private readonly Logger _logger;
+        private readonly string _logDocId;
+        private readonly string _orphanAttachmentsDocIdPrefix;
+        private readonly string _orphanCountersDocIdPrefix;
+        private readonly Slice _attachmentsSlice;
+        private readonly ByteStringContext _byteStringContext;
+        private readonly string _recoveryLogCollection;
+        private string GetOrphanAttachmentDocId(string hash) => $"{_orphanAttachmentsDocIdPrefix}/{hash}";
+
+        private RecoveredDatabaseCreator(DocumentDatabase documentDatabase, string recoverySession, Logger logger)
+        {
+            _logger = logger;
+            _recoveryLogCollection = $"RecoveryLog-{recoverySession}";
+            _orphanAttachmentsDocIdPrefix = $"OrphanAttachments/{recoverySession}";
+            _orphanCountersDocIdPrefix = $"OrphanCounters/{recoverySession}";
+            _logDocId = $"Log/{recoverySession}";
+            _database = documentDatabase;
+            _databaseDestination = new DatabaseDestination(documentDatabase);
+            _results = new SmugglerResult();
+            _databaseDestination.Initialize(new DatabaseSmugglerOptionsServerSide(), _results, ServerVersion.Build);
+            _contextDisposal = _database.DocumentsStorage.ContextPool.AllocateOperationContext(out _context);
+            _byteStringContext = new ByteStringContext(SharedMultipleUseFlag.None);
+            Slice.From(_byteStringContext, "Attachments", ByteStringType.Immutable, out _attachmentsSlice);
+
+            using (var tx = _context.OpenWriteTransaction())
+            {
+                using (var doc = _context.ReadObject(
```

Comment:
Wrap this in a method

Reply:
I am not following? ..

Reply:
Even in local function. So someone will know what you are doing here.